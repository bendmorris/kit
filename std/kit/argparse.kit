include "argp.h";

abstract ArgOption: argp_option {
    public static function new(key: Int): Self {
        var option = empty;
        option.key = key;
        return option;
    }

    public function long(name: CString): Self {
        this.name = name;
        return this;
    }

    public function help(doc: CString): Self {
        this.doc = doc;
        return this;
    }

    public function meta(meta: CString): Self {
        this.arg = meta;
        return this;
    }
}

abstract ArgParser[T]: argp {
    public static function new(options: Ptr[argp_option], parser: argp_parser_t, doc: CString): Self {
        var argParser = empty;
        argParser.options = options;
        argParser.parser = parser;
        argParser.doc = doc;
        return argParser;
    }

    public function parse(args: Slice[CString], state: T): T {
        var argIndex: Int = 0;
        argp_parse(this, args.length, args.data, ${ARGP_IN_ORDER: Int}, argIndex, &state);
        return state;
    }
}

abstract ArgState[T]: argp_state {
    public function state() {
        return this.input as Ptr[T];
    }
}
