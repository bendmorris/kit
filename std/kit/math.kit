var PI = 3.14159265358979323846264338327950288_f64;

function max[T: Numeric](x: T, y: T): T {
    return if y > x then y else x;
}

function min[T: Numeric](x: T, y: T): T {
    return if y < x then y else x;
}
