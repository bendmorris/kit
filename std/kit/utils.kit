// rules FormatString {
//     (fmt(${_: Byte})) => "%hh";
//     (fmt(${_: Short})) => "%h";
//     (fmt(${_: Int})) => "%l";
//     (fmt(${_: Long})) => "%ll";
//     (fmt(${_: Float})) => "%f";
//     (fmt(${_: Double})) => "%f";
// }

// public function println(s: CString) {
    // printf("%s\n", s);
// }

// public inline function print[S: ToString](s: S) {
    // println(s.toString());
// }

#[noreturn] function panic(msg: Const[CString], args...) {
    static if defined stderr {
        // use stderr in environments where it's defined
        vfprintf(stderr, msg, (args...));
        fputs("", stderr);
    } else {
        // fall back to stdout
        vprintf(msg, (args...));
        puts("");
    }
    abort();
}
