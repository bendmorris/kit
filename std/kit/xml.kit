import kit.sys.file;

include "libxml/xmlreader.h" => "xml2";

abstract Xml: Ptr[xmlDoc] {
    public static function parse(allocator: Box[Allocator], s: CString): Option[Xml] using implicit allocator {
        var result: Xml = xmlParseDoc(s as Const[Ptr[xmlChar]]) as Xml;
        return if result as Ptr[Void] == null then None else Some(result);
    }

    public static function parseFile(allocator: Box[Allocator], path: Const[CString]): Option[Xml] using implicit allocator {
        var result: Xml = xmlReadFile(path, null, 0) as Xml;
        return if result as Ptr[Void] == null then None else Some(result);
    }

    public function getRoot(): XmlNode {
        return xmlDocGetRootElement(this) as XmlNode;
    }

    public function free() {
        xmlFreeDoc(this);
    }
}

abstract XmlNode: Ptr[xmlNode] {
    public function getAttribute(name: CString): Option[XmlValue] {
        var result = xmlGetProp(this, name as Ptr[xmlChar]);
        return if result == null then None else Some(result as XmlValue);
    }

    public function getText(): Option[XmlValue] {
        var result = xmlNodeGetContent(this);
        return if result == null then None else Some(result as XmlValue);
    }

    rules {
        (for $child in $this {$e}) => {
            var node = $this;
            var $child: XmlNode = node.children as XmlNode;
            while $child as Ptr[Void] != null {
                {$e}
                $child = $child.next as XmlNode;
            }
        }
    }
}

abstract XmlValue: CString {
    public function free() {
        xmlFree(this);
    }
}
