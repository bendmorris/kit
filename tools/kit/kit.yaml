# Why are you building yourself? Why are you building yourself?
name: kit
version: 0.1.0
author: Ben Morris <ben@bendmorris.com>
url: https://www.kitlang.org
description: Build tool for the Kit programming language.

source:
- src

targets:
  release:
    executables:
    - mod: kit
      path: kit
    compiler_flags:
    - "-flto"
    - "-O3"
    - "-Os"
    linker_flags:
    - "-flto"

  debug:
    executables:
    - mod: kit
    compiler_flags:
    - "-g"

  test:
    executables:
    - mod: test

    source:
    - tests
