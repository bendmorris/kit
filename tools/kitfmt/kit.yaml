name: kitfmt
version: 0.1.0
author: Ben Morris <ben@bendmorris.com>
url: https://www.kitlang.org
description: Kit programming language code formatter.

source:
- src

targets:
  release:
    executables:
    - mod: kitfmt
    compiler_flags:
    - "-flto"
    - "-O3"
    - "-Os"

  debug:
    executables:
    - mod: kitfmt
    compiler_flags:
    - "-g"

  test:
    executables:
    - mod: test

    source:
    - tests
