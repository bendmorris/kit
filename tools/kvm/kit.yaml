name: kvm
version: 0.1.0
author: Ben Morris <ben@bendmorris.com>
url: https://www.kitlang.org
description: Kit compiler (kitc) version manager.

source:
- src

targets:
  release:
    executables:
    - mod: kvm
    compiler_flags:
    - "-flto"
    - "-O3"
    - "-Os"

  debug:
    executables:
    - mod: kvm
    compiler_flags:
    - "-g"

  test:
    executables:
    - mod: test

    source:
    - tests
