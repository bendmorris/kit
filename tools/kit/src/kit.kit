import kit.sys.file;
import kit.yaml;

import options;
import project;
import termcolor;

function main(argc: Int, argv: CArray[CString]) {
    printFox();

    var args = Slice.new(argv, argc);
    var options = KitOptions.parse(args);
    if options.subcommand.isNone() {
        // `kit` by itself will build & run the default target
        options.subcommand = Some(Run);
    }

    var target = options.target;
    var fp = File.read("kit.yaml");
    var yaml = Yaml.parseFile(fp).unwrap();
    var p = KitProject.fromYaml(yaml, target);
    var artifacts = p.getKitcArgs();
    for args in artifacts {
        options.addArgs(args);
        printf("$");
        for arg in args {
            printf(" %s", arg);
        }
        printf("\n");
        args.push(null);
        var status = execvp("kitc", args.data.data as Ptr[CString]);
        if status == -1 {
            Red.print("compilation has failed\n");
            panic("compilation has failed");
        } else {
            Green.print("success!\n");
        }
        args.free();
    }
}

function printFox() {
    Green.print("""
                       ,  ,
  _,-=._              /|_/|
  `-.}   `=._,.-=-._.,  """);
    Blue.print("@ @");
    Green.print("""._,
     `._ _,-.   )      _,.-'
        `    G.Q-^^u`u'
""");

    Blue.println("""
  oooo    oooo   o8o      ,
  `888   .8P'    `V'    ,o8
   888  d8'     oooo  .o888oo
   88888[       `888    888
   888`88b.      888    888
   888  `88b.    888    888 ,
  o888o  o888o  o888o   '888'
""");
}
