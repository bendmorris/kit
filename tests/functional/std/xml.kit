import kit.xml;

function testParse(s: CString) {
    var xml = Xml.parse(s);
    if !xml.isSome() {
        panic("failed to parse XML");
    }
    return xml.unwrap();
}

function main() {
    var doc1 = testParse("<test/>");
    doc1.free();

    var doc2 = testParse('<test a="1"><a b="1"></a><b></b>abc</test>');
    var root = doc2.getRoot();
    printf("%s\n", root.name);
    var prop = root.getAttribute("a");
    match prop {
        Some(v) => {
            printf("a=%s\n", v);
            v.free();
        }
        Nothing => {
            puts("no a attribute");
        }
    }
    prop = root.getAttribute("b");
    match prop {
        Nothing => {
            puts("no b attribute");
        }
    }
    for child in root {
        match child.type {
            XML_ELEMENT_NODE => {
                printf("element child: %s\n", child.name);
            }
            XML_TEXT_NODE => {
                var txt = child.getText().unwrap();
                printf("text child: %s\n", txt);
                txt.free();
            }
        }
    }
}
