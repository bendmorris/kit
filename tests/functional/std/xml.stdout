test
a=1
no b attribute
element child: a
element child: b
text child: abc
